opm/autodiff/BlackoilModelEbos.hpp

```diff
@@ -840,16 +833,18 @@ namespace Opm {
                 const double pvValue = ebosProblem.porosity(cell_idx) * ebosModel.dofTotalVolume( cell_idx );
                 pvSumLocal += pvValue;
 
-                for ( int phaseIdx = 0; phaseIdx < np; ++phaseIdx )
+                for (unsigned phaseIdx = 0; phaseIdx < FluidSystem::numPhases; ++phaseIdx)
```

Comment:
Is `FluidSystem::numPhases` equal to 2 or 3 for a two-phase case? To me this looks like we are changing from looping over "all active phases" to looping over "all possible phases" (and using `continue` to skip non-active ones). Is that correct?

Reply:
Yes.

Reply:
But is `FluidSystem::numPhases` always 3 for the black-oil case, even when (say) gas is not active?

Reply:
Yes. Number of components can be 2, 3 or 4.

Reply:
> But is FluidSystem::numPhases always 3 for the black-oil case, even when (say) gas is not active?

If it is the case, `numPhases` should be named something more specific and related to blackoil system. At least it should be obvious it is a constant 3 and not problem-dependent.

Reply:
It is problem dependent, but it is always 3 if you have a blackoil FluidSystem. Isn't the blackoil system 3 phase pr definition. The 2p versions of the model is more of a hack to be able to handle cases with a missing phase efficiently.

Reply:
I might understood wrongly. Under what kind of situations, `FluidSystem::numPhases` will not be 3?

Reply:
For instance 

> opm/material/fluidsystems/BrineCO2FluidSystem.hpp 

Not used by opm-flow, but used in ewoms.

Reply:
In the old implementation, `const int np = numPhases();` is telling the number of phases is problem dependent, while it is `const` for the specific model when the model is defined.  We used `maxNumPhases()` to tell the maximum possible number of phases for this system, which is also constant in some sense. Now the `numPhases` is closer to the old `maxNumPhases()` since even some phase is missing it stays the same in my opinion.

Reply:
Yes. 

You can also ask for FluidSystem::numActivePhases (this will be the old numPhases())

Reply:
```                 
                for (unsigned phaseIdx = 0; phaseIdx < FluidSystem::numPhases; ++phaseIdx)
                {
                    if (!FluidSystem::phaseIsActive(phaseIdx))
                        continue; 
 ```

Is it be same with 

```
                for (unsigned phaseIdx = 0; phaseIdx < FluidSystem::numActivePhases; ++phaseIdx)
                {
```

?
If yes, this can be a better way?

Reply:
Oh. They are not same. `phaseIdx` is in the range of `FluidSystem::numPhases`  while not in the range of `FluidSystem::numActivePhases` .

Reply:
One question here, if one phase `phaseIdx` is not active,  what is in ` fs.invB(phaseIdx)`. The loop gives the impression that even one phase `phaseIdx` is not active, `fs` still allocates some memory for that phase, is it correct?

Reply:
> if one phase `phaseIdx` is not active, what is in `fs.invB(phaseIdx)`

 that's a thing which the higher level code needs to decide. in ebos it currently is undefined in that case.